### ISSUE TITLE
[BUG] __pyx_fatalerror: Acquisition count is 0  (with memoryview slices, closures and exceptions)

### ISSUE THREAD

**Issue by navytux** [cython/cython#4798]
Hello up there. I've hit Python abort on a code that uses both memoryview slices and exception raised from inside closure. Please see below for details:

```cython
def f(s):
    g(s)

cdef g(const unsigned char[::1]s):
    cdef int i = 0
    cdef int l = len(s)
    def nextchar():
        nonlocal i
        if i >= l:
            raise ValueError("END")
        c = s[i]
        i += 1
        return c

    while 1:
        c = nextchar()
        print(c)

f(b'123')
```

```console
(neo) (py3.venv) (g.env) kirr@deca:~/tmp/trashme$ cythonize -i x.pyx && python -c 'import x'
Compiling /home/kirr/tmp/trashme/x.pyx because it changed.
[1/1] Cythonizing /home/kirr/tmp/trashme/x.pyx
/home/kirr/src/wendelin/venv/py3.venv/lib/python3.9/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /home/kirr/tmp/trashme/x.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)
running build_ext
building 'x' extension
creating /home/kirr/tmp/trashme/tmpapksgq8t/home
creating /home/kirr/tmp/trashme/tmpapksgq8t/home/kirr
creating /home/kirr/tmp/trashme/tmpapksgq8t/home/kirr/tmp
creating /home/kirr/tmp/trashme/tmpapksgq8t/home/kirr/tmp/trashme
x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/home/kirr/src/wendelin/venv/py3.venv/include -I/usr/include/python3.9 -c /home/kirr/tmp/trashme/x.c -o /home/kirr/tmp/trashme/tmpapksgq8t/home/kirr/tmp/trashme/x.o
x86_64-linux-gnu-gcc -pthread -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-z,relro -g -fwrapv -O2 -Wl,-z,relro -g -fwrapv -O2 -g -ffile-prefix-map=/build/python3.9-RNBry6/python3.9-3.9.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 /home/kirr/tmp/trashme/tmpapksgq8t/home/kirr/tmp/trashme/x.o -o /home/kirr/tmp/trashme/x.cpython-39-x86_64-linux-gnu.so

49
50
51
Fatal Python error: __pyx_fatalerror: Acquisition count is 0 (line 2297)
Python runtime state: initialized
Traceback (most recent call last):
  File "x.pyx", line 16, in x.g
    c = nextchar()
  File "x.pyx", line 10, in x.g.nextchar
    raise ValueError("END")
ValueError: END
Аварийный останов

```

**Expected behavior**
ValueError('END') raised but no `Fatal Python error: __pyx_fatalerror: Acquisition count is 0 (line 2297)`

**Environment (please complete the following information):**
 - OS: Debian GNU/Linux 11
 - Python version 3.9.2
 - Cython version 0.29.30

**Additional context**

https://github.com/cython/cython/issues/4662 might be related.

Thanks beforehand,
Kirill

**Comment by da-woods** [comment]
It's also broken on the current master version.

**Comment by navytux** [comment]
@da-woods, thanks for checking and confirming.

**Comment by da-woods** [comment]
FWIW the exception also doesn't make any difference. The memoryview isn't incremented when it's captured, but it's decremented when it's released, so it crashes whenever `nextchar` is cleaned up. I'm actually amazed that this has gone unreported for so long.

**Comment by navytux** [comment]
Thanks, @da-woods. Good point.

**Comment by scoder** [comment]
I wasn't sure either. I closed it because it's no longer an issue in the latest 0.29.x branch. But you're right, it's probably better to mark this as only (really) fixed in 3.0.

**Comment by navytux** [comment]
@da-woods, thanks for e4ef0c1e807aab8c20fb08b638550c912c166be3. And it is good to see https://github.com/cython/cython/pull/4848 is also there for Cython 3.0.
